ssm_jax/kmeans.py

```diff
@@ -0,0 +1,147 @@
+    key0, key1 = jr.split(key)
+
+    # Set the number of local seeding trials if none is given
+    if num_local_trials is None:
+        # This is what Arthur/Vassilvitskii tried, but did not report
+        # specific results for other than mentioning in the conclusion
+        # that it helped.
+        num_local_trials = 2 + jnp.log(num_clusters).astype(jnp.int32)
+
+    # Pick first center randomly and track index of point
+    initial_center = jr.choice(key0, X)[None, ...]
+
+    def euclidean_distance_square(x, y):
+        return jnp.square(x - y).sum(axis=-1)
+
+    # Initialize list of closest distances and calculate current potential
+    initial_distances = euclidean_distance_square(X, initial_center)
+
+    # Pick the remaining n_clusters-1 points
+    def find_center(distances, key):
+        current_pot = distances.sum()
+        random_values = jr.uniform(key, shape=(num_local_trials,)) * current_pot
+        candidate_ids = jnp.searchsorted(jnp.cumsum(distances), random_values)
+        # XXX: numerical imprecision can result in a candidate_id out of range
+        candidate_ids = jnp.clip(candidate_ids, a_max=num_samples - 1)
+
+        # Compute distances to center candidates
+        distance_to_candidates = vmap(lambda x, y: euclidean_distance_square(x[None, ...], y),
+                                      in_axes=(0, None))(X[candidate_ids], X)
+        # update closest distances squared and potential for each candidate
+        distance_to_candidates = vmap(jnp.minimum, in_axes=(0, None))(distance_to_candidates,
+                                                                      distances)
+
+        # Decide which candidate is the best
+        best_candidate = jnp.argmin(jnp.sum(distance_to_candidates, axis=-1))
+        distances = distance_to_candidates[best_candidate]
+        best_candidate = candidate_ids[best_candidate]
+
+        return distances, X[best_candidate]
+
+    if num_clusters == 1:
+        return initial_center
+    elif num_clusters == 2:
+        keys = jr.split(key1)[:1]
+    else:
+        keys = jr.split(key1, num_clusters - 1)
+
+    _, centers = lax.scan(find_center, initial_distances, keys)
+    centers = jnp.vstack([initial_center, centers])
+    return centers
+
+
+@partial(jit, static_argnums=(1,))
+def kmeans(X,
+           num_clusters,
+           max_iter=50,
+           threshold=1e-4,
+           initial_centroids=None,
+           num_init_iterations=1,
+           key=jr.PRNGKey(0)):
+    # https://colab.research.google.com/drive/1AwS4haUx6swF82w3nXr6QKhajdF8aSvA#scrollTo=XUaIhb7TmtGo
+    num_clusters = min(len(X), num_clusters)
+
+    def improve_centroids(state):
+        prev_centroids, prev_dists, _, i = state
+        assignment, distortions = find_closest_cluster(X, prev_centroids)
+
+        # Clip to change 0/0 later to 0/1
+        counts = jnp.clip(
+            (assignment[jnp.newaxis, :] == jnp.arange(num_clusters)[:, jnp.newaxis]).sum(
```

Comment:
I use `None` instead of `jnp.newaxis` to save characters. Some people don't like `None` because it's less explicit, but I think it's pretty standard now.

Reply:
Done